src/input/keyboard/mod.rs

```diff
@@ -249,12 +251,166 @@ impl<'a> KeysymHandle<'a> {
+                // be uniform across layouts.
+                if keysym
+                    .key_char()
+                    .map(|key| key.is_ascii_alphabetic())
+                    .unwrap_or(false)
+                {
+                    return Some(*keysym);
+                }
+            }
+        }
+
+        Some(base_sym)
+    }
+
     /// Returns the raw code in X keycode system (shifted by 8)
     pub fn raw_code(&'a self) -> Keycode {
         self.keycode
     }
 }
 
+impl<'a> XkbTypesHandle for KeysymHandle<'a> {
+    fn keymap(&self) -> &xkb::Keymap {
+        self.keymap
+    }
+
+    fn state(&self) -> &xkb::State {
+        self.state
+    }
+}
+
+/// The currently active state of the Xkb.
+pub struct XkbContext<'a> {
+    state: &'a mut xkb::State,
+    keymap: &'a mut xkb::Keymap,
+    mods_state: &'a mut ModifiersState,
+    mods_changed: &'a mut bool,
+}
+
+impl<'a> XkbContext<'a> {
+    /// Set layout of the keyboard to the given index.
+    ///
+    /// Returns `true` when the layout got applied.
+    pub fn set_layout(&mut self, layout: Layout) {
+        let state = self.state.update_mask(
+            self.mods_state.serialized.depressed,
+            self.mods_state.serialized.latched,
+            self.mods_state.serialized.locked,
+            0,
+            0,
+            layout.0,
+        );
+
+        if state != 0 {
+            self.mods_state.update_with(self.state);
+            *self.mods_changed = true;
+        }
+    }
+
+    /// Switches layout forward cycling when cyclying when it reaches the end.
+    pub fn cycle_next_layout(&mut self) {
+        let next_layout = (self.active_layout().0 + 1) % self.keymap.num_layouts();
+        self.set_layout(Layout(next_layout));
+    }
+
+    /// Switches layout backward cycling when cyclying when it reaches the start.
+    pub fn cycle_prev_layout(&mut self) {
+        let num_layouts = self.keymap.num_layouts();
+        let next_layout = (num_layouts + self.active_layout().0 - 1) % num_layouts;
+        self.set_layout(Layout(next_layout));
+    }
+}
+
+impl<'a> fmt::Debug for XkbContext<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XkbContext")
+            .field("mods_state", &self.mods_state)
+            .field("mods_changed", &self.mods_changed)
+            .finish()
+    }
+}
+
+impl<'a> XkbTypesHandle for XkbContext<'a> {
+    fn keymap(&self) -> &xkb::Keymap {
+        self.keymap
+    }
+
+    fn state(&self) -> &xkb::State {
+        self.state
+    }
+}
+
+/// Query and manipulate Xkb layouts.
+pub trait XkbContextHandler: XkbTypesHandle {
+    /// Get the active layout of the keyboard.
+    fn active_layout(&self) -> Layout {
+        (0..self.keymap().num_layouts())
+            .find(|&idx| {
+                self.state()
+                    .layout_index_is_active(idx, XKB_STATE_LAYOUT_EFFECTIVE)
+            })
+            .map(Layout)
+            .unwrap_or_default()
```

Comment:
Does `unwrap_or_default()` make sense here?

Reply:
yes, default is base layout.